dynamodb/schema.go

```diff
@@ -0,0 +1,358 @@
+	}
+
+	// Primary keys
+	for _, i := range result.Table.KeySchema {
+		s.PrimaryKeys = append(s.PrimaryKeys, *i.AttributeName)
+	}
+
+	// Secondary indexes
+	for _, i := range result.Table.GlobalSecondaryIndexes {
+		var keys []string
+		for _, j := range i.KeySchema {
+			keys = append(keys, *j.AttributeName)
+		}
+		s.SecIndexes = append(s.SecIndexes, index{Name: *i.IndexName, Keys: keys})
+	}
+
+	return nil
+}
+
+func (s *dynamoDBSchema) scanSampleData(client dynamoClient, sampleSize int64) (map[string]map[string]int64, int64, error) {
+	// A map from column name to a count map of possible data types.
+	stats := make(map[string]map[string]int64)
+	var lastEvaluatedKey map[string]*dynamodb.AttributeValue
+	var count int64
+
+	for {
+		// Build the query input parameters
+		params := &dynamodb.ScanInput{
+			TableName: aws.String(s.TableName),
+		}
+		if lastEvaluatedKey != nil {
+			params.ExclusiveStartKey = lastEvaluatedKey
+		}
+
+		// Make the DynamoDB Query API call
+		result, err := client.Scan(params)
+		if err != nil {
+			return nil, 0, fmt.Errorf("failed to make Query API call for table %v: %v", s.TableName, err)
+		}
+
+		// Iterate the items returned
+		for _, attrsMap := range result.Items {
+			for attrName, attr := range attrsMap {
+				if _, ok := stats[attrName]; !ok {
+					stats[attrName] = make(map[string]int64)
+				}
+				incDyDataTypeCount(attrName, attr, stats[attrName])
+			}
+
+			count++
+			if count >= sampleSize {
+				return stats, count, nil
+			}
+		}
+		if result.LastEvaluatedKey == nil {
+			break
+		}
+		// If there are more rows, then continue.
+		lastEvaluatedKey = result.LastEvaluatedKey
+	}
+	return stats, count, nil
+}
+
+func incDyDataTypeCount(attrName string, attr *dynamodb.AttributeValue, s map[string]int64) {
+	if attr.S != nil {
+		incCount(s, typeString)
+	} else if attr.BOOL != nil {
+		incCount(s, typeBool)
+	} else if attr.N != nil {
+		if int64Parsable(*attr.N) {
+			incCount(s, typeNumberInt)
+		} else {
+			incCount(s, typeNumberFloat)
+		}
+	} else if len(attr.B) != 0 {
+		incCount(s, typeBinary)
+	} else if attr.NULL != nil {
+		// Skip, if not present, it means nullable.
+	} else if len(attr.L) != 0 {
+		incCount(s, typeList)
+	} else if len(attr.M) != 0 {
+		incCount(s, typeMap)
+	} else if len(attr.SS) != 0 {
+		incCount(s, typeStringSet)
+	} else if len(attr.NS) != 0 {
+		if int64Parsable(*attr.NS[0]) {
+			incCount(s, typeNumberIntSet)
+		} else {
+			incCount(s, typeNumberFloatSet)
+		}
+	} else if len(attr.BS) != 0 {
+		incCount(s, typeBinarySet)
+	} else {
+		log.Printf("Invalid DynamoDB data type: %v - %v", attrName, attr)
+	}
+}
+
+type statItem struct {
+	Type  string
+	Count int64
+}
+
+func (s *dynamoDBSchema) inferDataTypes(stats map[string]map[string]int64, rows int64) {
+	if s.ColumnTypes == nil {
+		s.ColumnTypes = make(map[string]colType)
+	}
+
+	for col, countMap := range stats {
+		var statItems, candidates []statItem
+		var presentRows, normRows int64
```

Comment:
What's the reason to distinguish normRows and presentRows?
They can only differ by a very tiny amount.
Or am I missing some corner case?

Reply:
Maybe the name is confusing. 

`rows`: the total number of sample rows. 
`presentRows`: the number of rows that have a value on this column. 
`rows-presentRows`: the number of rows that do not have a value on this column. It used to check whether or not this column can be nullable. 

`normRows`: the number of rows after removing erroneous ones (the percent of type is <= err threshold) and nullable rows. 

I will use an example to illustrate this: 

```bash
# assume error threshold = 0.1%
rows = 1000

stats = {
    "Username": {
        "String":    800,
        "NumberInt": 1,
        "Bytes": 1,
        "Bool": 1,
    },
}
```

Then, 

```
presentRows = 800 + 1 + 1 + 1 = 803
rows-presentRows = 1000 - 803 = 197 > 1  # This means that this column should be nullable. 

normRows = 800
```

If you think `normRows` and `presentRows` only have a very tiny difference, we can just use `presentRows`.

Reply:
`presentRows` is tested by the test case: 

```bash
		"enough_null_row": {
			"NumberInt": 900,
		},
```

`normRows` is tested by the test case: 

```bash
		"equal_conflict_row_after_norm": {
			"String":    400,
			"NumberInt": 400,
		},
```

Reply:
What puzzled me as I was looking over the code was why there were two counts. And it made the code harder to understand since there were two very similar concepts to keep track of.

At line 253, where we check the error threshold, the comment explains that there is a high chance they are mistakes and that we should discard them.

We then include them for the purposes of determining nullable, but ignore them for determining types.

If we justify ignoring them because they are likely mistakes, shouldn't we ignore them for both calculations?

I think either we should be consistent, or we should have a good reason (and a comment) to explain why we are doing something different for these two cases.

Reply:
> If we justify ignoring them because they are likely mistakes, shouldn't we ignore them for both calculations?

Yes, we can. But in this case, we still need to keep `presentRows`, which is used to calculate how many rows are not given. If we ignore discarded rows when determining nullable, then the code would become: 

```go
nullable := float64(rows-presentRows)/float64(normRows) > errThreshold
```

We changed the denominator from `rows` to `normRows`. 

The simple change would be to remove `normRows` so we always use `rows` instead of `normRows`.